src/i2c.rs

```diff
@@ -7,186 +7,326 @@
+            expected_data: Vec::new(),
+            response_data: response,
+        }
+    }
+
+    /// Create a WriteRead transaction
+    pub fn write_read(addr: u8, expected: Vec<u8>, response: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::WriteRead,
+            expected_addr: addr,
+            expected_data: expected,
+            response_data: response,
+        }
+    }
+}
+
+/// Mock I2C implementation
+/// This supports the specification and evaluation of expectations to allow automated testing of I2C based drivers.
+/// Mismatches between expectations will cause runtime assertions to assist in locating the source of the fault.
+pub struct Mock {
+    expected: Arc<Mutex<VecDeque<Transaction>>>,
 }
 
-impl<'a> Mock<'a> {
+impl Mock {
     pub fn new() -> Self {
         Mock {
-            data: &[],
-            address: None,
-            buf: [0; WRITE_BUF_SIZE],
-            buf_bytes_written: 0,
+            expected: Arc::new(Mutex::new(VecDeque::new())),
         }
     }
 
-    /// Set data that will be read by `read()`.
-    pub fn set_read_data(&mut self, data: &'a [u8]) {
-        self.data = data;
+    /// Set expectations on the I2C interface
+    /// This is a list of I2C transactions to be executed in order
+    /// Note that setting this will overwrite any existing expectations
+    pub fn expect(&mut self, expected: Vec<Transaction>) {
+        let mut e = self.expected.lock().unwrap();
+        *e = expected.into();
     }
 
-    /// Return the data that was written by the last write command.
-    pub fn get_write_data(&self) -> &[u8] {
-        &self.buf[0..self.buf_bytes_written]
+    /// Assert that all expectations on a given Mock have been met
+    pub fn done(&mut self) {
+        let expected = self.expected.lock().unwrap();
+        assert_eq!(expected.len(), 0);
     }
+}
 
-    /// Return the address that was used by the last read or write command.
-    pub fn get_last_address(&self) -> Option<u8> {
-        self.address
+impl Clone for Mock {
+    fn clone(&self) -> Mock {
+        Mock{expected: self.expected.clone()}
     }
 }
 
-impl<'a> i2c::Read for Mock<'a> {
+impl i2c::Read for Mock {
     type Error = MockError;
 
-    fn read(&mut self, address: u8, mut buffer: &mut [u8]) -> Result<(), Self::Error> {
-        self.address = Some(address);
-        self.data.read(&mut buffer)?;
+    fn read(&mut self, address: u8, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        let w = self
+            .expected
+            .lock()
+            .unwrap()
+            .pop_front()
+            .expect("no pending expectation for i2c::read call");
+
+        assert_eq!(w.expected_mode, Mode::Read);
+        assert_eq!(w.expected_addr, address);
+
+        assert_eq!(buffer.len(), w.response_data.len(), "mismatched response length for i2c:read");
+        for i in 0..buffer.len() {
+            buffer[i] = w.response_data[i];
+        }
```

Comment:
I think you can use `buffer.copy_from_slice(&w.response_data)` instead of the loop. Can probably be applied in the SPI code too. Also in the `WriteRead` implementation.

Reply:
Yep, can do